### DISCUSSION TITLE
MinimalApi : Using [AsParameters] to inject a Handler

### DISCUSSION THREAD

**Discussion by srollinet** [dotnet/aspnetcore#56809] [Q&A]
Hi!

I've been experimenting with Minimal APIs to find a clean way to organize my endpoints, and I finally came across something I am quite happy with. I'd like to get your opinion on it.

The idea is to use `[AsParameters]` to inject a handler object that will hold everything needed to process the request (parameters, services, etc.), as well as a `Handle()` method that will be used as the handler delegate. Of course, the class could also contain private methods if the logic is more complexe.

A simple example looks like this
Note: I included some validation and cancellation support to illustrate what can be done.

```cs
public class CreateTodoListHandler
{
    [FromBody]
    public required TodoListCreateDto CreateDto { get; init; }
    
    public required TodoListCreateDtoValidator Validator { get; init; }
    
    public required ApplicationDbContext Context { get; init; }
    
    public required CancellationToken CancellationToken { get; init; }

    public async Task<Results<Ok<int>, ValidationProblem>> Handle()
    {
        var validationResult = await Validator.ValidateAsync(CreateDto, CancellationToken);
        if (!validationResult.IsValid)
        {
            return TypedResults.ValidationProblem(validationResult.ToDictionary());
        }
        
        var list = new TodoList {Title = CreateDto.Title};
        Context.TodoLists.Add(list);
        await Context.SaveChangesAsync(CancellationToken);
            
        return TypedResults.Ok(list.Id);
    }
}
```

With the mapping code

```cs
public static class EndpointRouteBuilderExtensions
{
    public static void MapTodoListEndpoints(this IEndpointRouteBuilder endpoints)
    {
        var group = endpoints.MapGroup("api/v1/TodoLists");

        group.MapPost("", ([AsParameters] CreateTodoListHandler handler) => handler.Handle());
        group.MapPost("{ListId}/Items", ([AsParameters] AddTodoItemHandler handler) => handler.Handle());
        // ........
    }
}
```

Is this a recommended pattern? Does using `[AsParameters]` for handler injection align with best practices for .NET Minimal APIs?

Are there alternative patterns or approaches you would recommend for achieving similar goals in a more efficient or standardized way?

Thanks a lot for your answers 😄

**Comment by nquandt** [comment]
I'm not sure if this is similar to what you are trying to accomplish but I saw your discussion and thought I'd share: https://github.com/dotnet/aspnetcore/discussions/60194

**Reply by srollinet** [reply]
Hey, thanks for the sharing! I also tried something like that long time ago but I gave up.

Finally I found my own preferred way to structure minimal endpoint inspired by this example: https://github.com/jonowilliams26/StructuredMinimalApi/blob/master/src/Users/Endpoints/GetUserComments.cs

Minimal code with simple interfaces and extension methods without black magic :)
